src/Mvc/Mvc.Core/src/Routing/UrlHelperBase.cs

```diff
@@ -43,45 +43,43 @@ protected UrlHelperBase(ActionContext actionContext)
         /// <inheritdoc />
         public virtual bool IsLocalUrl(string url)
         {
-            if (string.IsNullOrEmpty(url))
+            ReadOnlySpan<char> span = url;
+
+            if (span.IsEmpty)
             {
                 return false;
             }
 
-            // Allows "/" or "/foo" but not "//" or "/\".
-            if (url[0] == '/')
+            if (span[0] == '~')
             {
-                // url is exactly "/"
-                if (url.Length == 1)
-                {
-                    return true;
-                }
-
-                // url doesn't start with "//" or "/\"
-                if (url[1] != '/' && url[1] != '\\')
-                {
-                    return true;
-                }
+                span = span[1..];
+            }
 
+            // url does not start with "/" or "~/"
+            if (span[0] != '/')
+            {
                 return false;
             }
 
-            // Allows "~/" or "~/foo" but not "~//" or "~/\".
-            if (url[0] == '~' && url.Length > 1 && url[1] == '/')
+            // url is exactly "/" or "~/"
+            if ((uint)span.Length <= 1)     // PERF: elides the bounds-check below
             {
-                // url is exactly "~/"
-                if (url.Length == 2)
-                {
-                    return true;
-                }
+                return true;
+            }
 
-                // url doesn't start with "~//" or "~/\"
-                if (url[2] != '/' && url[2] != '\\')
+            // url doesn't start with "//" or "/\"
+            if (span[1] != '/' && span[1] != '\\')
+            {
+                // check that url does not contain any (ASCII) control character
+                for (var i = 1; i < span.Length; ++i)
                 {
-                    return true;
+                    if (span[i] < 0x20 || span[i] == 0x7F)
```

Comment:
Why not use `char.IsControl(span[i])`?

Reply:
@GrabYourPitchforks plans on porting some of the perf improvements made to https://source.dot.net/#System.Private.CoreLib/Rune.cs,1132 to `Char.IsControl`. So it should be just as fast by .NET 5.0 👍

Reply:
Yeah, if you're trying to block control characters just rely on `char.IsControl`. That way you don't need to spend weeks upon weeks poring over ISO and Unicode specifications. Let us do it for you. :)

Reply:
Good to know -- I'll update this.